builder/dockerfile/instructions/commands.go

```diff
@@ -0,0 +1,297 @@
+func (c *LabelCommand) Expand(expander SingleWordExpander) error {
+	return expandKvpsInPlace(c.Labels, expander)
+}
+
+type AddCommand struct {
+	BaseCommand
+	Srcs []string
+	Dest string
+}
+
+func (c *AddCommand) Expand(expander SingleWordExpander) error {
+	dst, err := expander(c.Dest)
+	if err != nil {
+		return err
+	}
+	c.Dest = dst
+	return expandSliceInPlace(c.Srcs, expander)
+}
+
+type CopyCommand struct {
+	BaseCommand
+	Srcs []string
+	Dest string
+	From string
+}
+
+func (c *CopyCommand) Expand(expander SingleWordExpander) error {
+	dst, err := expander(c.Dest)
+	if err != nil {
+		return err
+	}
+	c.Dest = dst
+	from, err := expander(c.From)
+	if err != nil {
+		return err
+	}
+	c.From = from
+	return expandSliceInPlace(c.Srcs, expander)
+}
+
+type FromCommand struct {
+	BaseCommand
+	BaseName  string
+	StageName string
+}
+
+type OnbuildCommand struct {
+	BaseCommand
+	Expression string
+}
+
+func (c *OnbuildCommand) Expand(expander SingleWordExpander) error {
+	p, err := expander(c.Expression)
+	if err != nil {
+		return err
+	}
+	c.Expression = p
+	return nil
+}
+
+type WorkdirCommand struct {
+	BaseCommand
+	Path string
+}
+
+func (c *WorkdirCommand) Expand(expander SingleWordExpander) error {
+	p, err := expander(c.Path)
+	if err != nil {
+		return err
+	}
+	c.Path = p
+	return nil
+}
+
+type RunCommand struct {
+	BaseCommand
+	Expression   strslice.StrSlice
+	PrependShell bool
+}
+
+type CmdCommand struct {
+	BaseCommand
+	Cmd          strslice.StrSlice
+	PrependShell bool
+}
+
+type HealthCheckCommand struct {
+	BaseCommand
+	Health *container.HealthConfig
+}
+
+type EntrypointCommand struct {
+	BaseCommand
+	Cmd          strslice.StrSlice
+	PrependShell bool
+}
+
+type ExposeCommand struct {
+	BaseCommand
+	Ports []string
+}
+
+type UserCommand struct {
+	BaseCommand
+	User string
+}
+
+func (c *UserCommand) Expand(expander SingleWordExpander) error {
+	p, err := expander(c.User)
+	if err != nil {
+		return err
+	}
+	c.User = p
+	return nil
+}
+
+type VolumeCommand struct {
+	BaseCommand
+	Volumes []string
+}
+
+func (c *VolumeCommand) Expand(expander SingleWordExpander) error {
+	return expandSliceInPlace(c.Volumes, expander)
+}
+
+type StopSignalCommand struct {
+	BaseCommand
+	Signal string
+}
+
+func (c *StopSignalCommand) Expand(expander SingleWordExpander) error {
+	p, err := expander(c.Signal)
+	if err != nil {
+		return err
+	}
+	c.Signal = p
+	return nil
+}
+
+type ArgCommand struct {
+	BaseCommand
+	Name  string
+	Value *string
+}
+
+func (c *ArgCommand) Expand(expander SingleWordExpander) error {
+	p, err := expander(c.Name)
+	if err != nil {
+		return err
+	}
+	c.Name = p
+	if c.Value != nil {
+		p, err = expander(*c.Value)
+		if err != nil {
+			return err
+		}
+		c.Value = &p
+	}
+	return nil
+}
+
+type ShellCommand struct {
+	BaseCommand
+	Shell strslice.StrSlice
+}
+
+type ResumeBuildCommand struct {
+	BaseConfig *container.Config
+}
+
+type BuildableStage struct {
+	Name     string
+	Commands []interface{}
```

Comment:
There is one specific command that does not embed CommandBase as it is not parsed from source code (ResumeBuild, used in BuildFromConfig). This also opens up some cool features like in the future, being able to split typed commands into more fine grained subcommands to increase parallelism opportunities. (for exemple, having the AddCommand split into many `fetch` subcommands -1 per source- and join them in a `mergeInto` command)